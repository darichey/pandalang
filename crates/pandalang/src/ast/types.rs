// TODO: syntax for polymorphic types
#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Simple(String),
    Fun(Fun),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub from: Box<Type>,
    pub to: Box<Type>,
}
