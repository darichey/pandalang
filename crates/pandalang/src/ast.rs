#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub patt: Pattern,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub patt: Pattern,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
        }
        .to_string()
    }
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Pattern {
    Id {
        name: String,
        // typ: Type
    },
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Base { name: String },
    Fun(Box<Type>, Box<Type>),
}
