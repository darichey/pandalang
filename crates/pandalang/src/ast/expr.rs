#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Unit,
    Bool(Bool),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
    If(If),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Bool {
    pub b: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub arg: String,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct If {
    pub check: Box<Expr>,
    pub then: Box<Expr>,
    pub els: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eql,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
            BinOpKind::Eql => "==",
        }
        .to_string()
    }
}
