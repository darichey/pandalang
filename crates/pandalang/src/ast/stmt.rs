use super::{expr::Expr, types::Type};

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Stmt {
    Let(Let),
    Declare(Declare),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub rec: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Declare {
    pub name: String,
    pub typ: Type,
}
