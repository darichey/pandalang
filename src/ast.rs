#[derive(PartialEq, Eq, Debug, Clone)]
pub enum ExprF<A> {
    Int {
        n: i64,
    },
    Var {
        name: String,
    },
    BinOp {
        left: A,
        right: A,
        kind: BinOpKind,
    },
    Fun{
        arg: String,
        body: A,
    },
    App{
        fun: A,
        arg: A,
    },
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Expr(pub Box<ExprF<Expr>>);

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}
