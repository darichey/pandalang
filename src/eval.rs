use std::collections::HashMap;

use crate::ast::{BinOpKind, Expr, ExprF};

pub struct Env {
    pub bindings: HashMap<String, Expr>,
}

macro_rules! new_env {
    ($($k:expr => $v:expr),* $(,)?) => {{
        Env { bindings: std::collections::HashMap::from([$(($k.to_string(), Expr(Box::new(ExprF::Int { n: $v }))),)*]) }
    }};
}

pub(crate) use new_env;

impl Env {
    pub fn lookup(&self, x: &String) -> Expr {
        self.bindings.get(x).unwrap().clone()
    }
}

pub fn eval(expr: Expr, env: &Env) -> Expr {
    match *expr.0 {
        ExprF::Int { n } => expr,
        ExprF::Var { name } => env.lookup(&name),
        ExprF::BinOp { left, right, kind } => {
            let f = match kind {
                BinOpKind::Add => std::ops::Add::add,
                BinOpKind::Sub => std::ops::Sub::sub,
                BinOpKind::Mul => std::ops::Mul::mul,
                BinOpKind::Div => std::ops::Div::div,
            };

            let (x, y) = match (*eval(left, env).0, *eval(right, env).0) {
                (ExprF::Int { n: x }, ExprF::Int { n: y }) => (x, y),
                _ => panic!("oh god oh fuck"),
            };

            Expr(Box::new(ExprF::Int { n: f(x, y) }))
        }
        ExprF::Fun { arg, body } => Expr(Box::new(ExprF::Fun { arg, body })),
        ExprF::App { fun, arg } => {
            let (arg_name, body) = match *eval(fun, env).0 {
                ExprF::Fun { arg, body } => (arg, body),
                _ => panic!("oh god oh fuck"),
            };
            let arg = eval(arg, env);

            eval(cas(body, arg, arg_name), env)
        }
    }
}

fn cas(e1: Expr, e2: Expr, var: String) -> Expr {
    return match *e1.0 {
        ExprF::Var { name } => {
            if var == name {
                e2
            } else {
                Expr(Box::new(ExprF::Var { name }))
            }
        }
        ExprF::Fun { arg, body } => {
            if var == arg {
                Expr(Box::new(ExprF::Fun { arg, body }))
            } else {
                Expr(Box::new(ExprF::Fun {
                    arg,
                    body: cas(body, e2, var),
                }))
            }
        }
        ExprF::App { fun, arg } => Expr(Box::new(ExprF::App {
            fun: cas(fun, e2.clone(), var.clone()),
            arg: cas(arg, e2, var),
        })),
        ExprF::Int { n } => Expr(Box::new(ExprF::Int { n })),
        ExprF::BinOp { left, right, kind } => Expr(Box::new(ExprF::BinOp {
            left: cas(left, e2.clone(), var.clone()),
            right: cas(right, e2, var),
            kind,
        })),
    };
}

#[cfg(test)]
mod tests {
    use super::Env;
    use crate::{ast::Expr, ast::ExprF, eval, parser};

    fn eval_test(s: String) -> Expr {
        eval::eval(
            parser::parse(s.as_str()).unwrap(),
            &new_env!("x" => 0, "y" => 1, "x'" => 2, "foo" => 3, "a" => 4, "b" => 5, "c" => 6, "d" => 7, "e" => 8),
        )
    }

    #[test]
    fn evals() {
        insta::glob!("snapshot_inputs/**/*.panda", |path| {
            let source = eval_test(std::fs::read_to_string(&path).unwrap());
            insta::assert_debug_snapshot!(source);
        });
    }
}
