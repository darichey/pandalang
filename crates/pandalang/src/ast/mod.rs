use self::stmt::Stmt;

pub mod expr;
pub mod stmt;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
