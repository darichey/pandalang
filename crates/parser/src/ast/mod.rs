use self::stmt::Stmt;

pub mod expr;
pub mod stmt;
pub mod types;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
