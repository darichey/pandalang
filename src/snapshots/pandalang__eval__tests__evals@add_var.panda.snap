---
source: src/eval.rs
expression: source
input_file: src/snapshot_inputs/bin_op/add_var.panda
---
Expr(
    Int {
        n: 9,
    },
)
